File: backend/scripts/simulate_telemetry.js
#!/usr/bin/env node
/**
 * FPSaver — Create devices + Insert simulated telemetry (24h)
 * Run: node scripts/simulate_telemetry.js
 */
const db = require('../src/config/database');
const { hash } = require('../src/utils/password');

async function main() {
    const client = await db.getClient();

    try {
        // 1. Get factory
        console.log('[1] Getting factory...');
        let factory = (await client.query('SELECT * FROM factories LIMIT 1')).rows[0];
        if (!factory) {
            // Get company
            const company = (await client.query('SELECT id FROM companies LIMIT 1')).rows[0];
            if (!company) { console.error('No companies exist!'); process.exit(1); }
            const fRes = await client.query(
                `INSERT INTO factories (company_id, name, location_address, city, latitude, longitude, timezone, mqtt_topic)
                 VALUES ($1, 'Planta Bilbao Norte', 'Polígono Industrial Asúa, 12', 'Bilbao', 43.263, -2.935, 'Europe/Madrid', null) RETURNING *`,
                [company.id]
            );
            factory = fRes.rows[0];
            await client.query('UPDATE factories SET mqtt_topic = $1 WHERE id = $2', [`factory/${factory.id}`, factory.id]);
            console.log(`   Created factory: ${factory.name}`);
        } else {
            console.log(`   Factory: ${factory.name} (${factory.id.substring(0, 8)}...)`);
        }

        // 2. Get or create devices
        console.log('[2] Checking devices...');
        let devices = (await client.query('SELECT * FROM devices WHERE factory_id = $1', [factory.id])).rows;

        if (devices.length === 0) {
            console.log('   Creating 4 devices...');
            const devicesToCreate = [
                { name: 'Compresor 1', type: 'trifasica', addr: 1, model: 'EM340', sn: 'SN-CG-001' },
                { name: 'Compresor 2', type: 'trifasica', addr: 2, model: 'EM340', sn: 'SN-CG-002' },
                { name: 'Horno Industrial', type: 'trifasica', addr: 3, model: 'EM340', sn: 'SN-CG-003' },
                { name: 'Iluminación Nave', type: 'monofasica', addr: 4, model: 'EM111', sn: 'SN-CG-004' },
            ];
            for (const d of devicesToCreate) {
                const r = await client.query(
                    `INSERT INTO devices (factory_id, name, device_type, modbus_address, model, serial_number)
                     VALUES ($1, $2, $3, $4, $5, $6) RETURNING *`,
                    [factory.id, d.name, d.type, d.addr, d.model, d.sn]
                );
                devices.push(r.rows[0]);
                console.log(`   ✓ ${d.name}`);
            }
        } else {
            console.log(`   ${devices.length} devices exist`);
        }

        // 3. Insert telemetry
        console.log('[3] Inserting 24h of simulated telemetry...');
        const now = new Date();
        let totalInserted = 0;

        await client.query('BEGIN');

        for (const device of devices) {
            const isTri = device.device_type === 'trifasica';
            let deviceReadings = 0;

            for (let i = 0; i < 288; i++) { // 24h * 12/h = 288 readings
                const ts = new Date(now.getTime() - i * 5 * 60 * 1000);
                const h = ts.getUTCHours();
                const loadFactor = (h >= 6 && h < 22)
                    ? 0.6 + 0.3 * Math.sin(Math.PI * (h - 6) / 16)
                    : 0.15 + 0.05 * Math.random();

                const gauss = () => { let u = 0, v = 0; while (!u) u = Math.random(); while (!v) v = Math.random(); return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v); };

                let data;
                if (isTri) {
                    const bp = (3000 + Math.random() * 9000) * loadFactor;
                    const v1 = 228 + gauss() * 2, v2 = 229 + gauss() * 2, v3 = 231 + gauss() * 2;
                    const pf = Math.min(0.98, Math.max(0.78, 0.91 + gauss() * 0.03));
                    const i1 = bp / (v1 * pf) * (1 + gauss() * 0.05);
                    const i2 = bp / (v2 * pf) * (1 + gauss() * 0.05);
                    const i3 = bp / (v3 * pf) * (1 + gauss() * 0.05);
                    const sinPF = Math.sin(Math.acos(pf));

                    data = {
                        voltage_l1_n: +v1.toFixed(1), voltage_l2_n: +v2.toFixed(1), voltage_l3_n: +v3.toFixed(1),
                        current_l1: +i1.toFixed(2), current_l2: +i2.toFixed(2), current_l3: +i3.toFixed(2),
                        power_w_l1: +(v1 * i1 * pf).toFixed(1), power_w_l2: +(v2 * i2 * pf).toFixed(1), power_w_l3: +(v3 * i3 * pf).toFixed(1),
                        power_w_total: +(v1 * i1 * pf + v2 * i2 * pf + v3 * i3 * pf).toFixed(1),
                        power_va_l1: +(v1 * i1).toFixed(1), power_va_l2: +(v2 * i2).toFixed(1), power_va_l3: +(v3 * i3).toFixed(1),
                        power_var_l1: +(v1 * i1 * sinPF).toFixed(1), power_var_l2: +(v2 * i2 * sinPF).toFixed(1), power_var_l3: +(v3 * i3 * sinPF).toFixed(1),
                        power_factor_l1: +pf.toFixed(3), power_factor_l2: +pf.toFixed(3), power_factor_l3: +pf.toFixed(3),
                        power_factor: +pf.toFixed(3),
                        frequency_hz: +(50 + gauss() * 0.05).toFixed(2),
                        energy_kwh_total: +(1000 + Math.random() * 49000).toFixed(1),
                    };
                } else {
                    const bp = (500 + Math.random() * 2000) * loadFactor;
                    const v = 230 + gauss() * 2;
                    const pf = Math.min(0.99, Math.max(0.85, 0.95 + gauss() * 0.02));
                    const ci = bp / (v * pf);
                    const sinPF = Math.sin(Math.acos(pf));

                    data = {
                        voltage_l1_n: +v.toFixed(1),
                        current_l1: +ci.toFixed(2),
                        power_w_l1: +bp.toFixed(1), power_w_total: +bp.toFixed(1),
                        power_va_l1: +(v * ci).toFixed(1),
                        power_var_l1: +(v * ci * sinPF).toFixed(1),
                        power_factor_l1: +pf.toFixed(3), power_factor: +pf.toFixed(3),
                        frequency_hz: +(50 + gauss() * 0.05).toFixed(2),
                        energy_kwh_total: +(100 + Math.random() * 4900).toFixed(1),
                    };
                }

                await client.query(
                    `INSERT INTO telemetry (
                        time, device_id, device_type,
                        voltage_l1_n, voltage_l2_n, voltage_l3_n,
                        current_l1, current_l2, current_l3,
                        power_w_l1, power_w_l2, power_w_l3, power_w_total,
                        power_va_l1, power_va_l2, power_va_l3,
                        power_var_l1, power_var_l2, power_var_l3,
                        power_factor, power_factor_l1, power_factor_l2, power_factor_l3,
                        frequency_hz, energy_kwh_total, raw_data
                    ) VALUES (
                        $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13,
                        $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26
                    )`,
                    [
                        ts.toISOString(), device.id, device.device_type,
                        data.voltage_l1_n || null, data.voltage_l2_n || null, data.voltage_l3_n || null,
                        data.current_l1 || null, data.current_l2 || null, data.current_l3 || null,
                        data.power_w_l1 || null, data.power_w_l2 || null, data.power_w_l3 || null, data.power_w_total || null,
                        data.power_va_l1 || null, data.power_va_l2 || null, data.power_va_l3 || null,
                        data.power_var_l1 || null, data.power_var_l2 || null, data.power_var_l3 || null,
                        data.power_factor || null, data.power_factor_l1 || null, data.power_factor_l2 || null, data.power_factor_l3 || null,
                        data.frequency_hz || null, data.energy_kwh_total || null, JSON.stringify(data)
                    ]
                );

                deviceReadings++;
                totalInserted++;
            }

            // Update realtime cache with latest reading
            const latestData = {
                voltage_l1_n: isTri ? 229.5 : 230.1,
                voltage_l2_n: isTri ? 230.2 : null,
                voltage_l3_n: isTri ? 231.0 : null,
                current_l1: isTri ? 28.5 : 8.2,
                current_l2: isTri ? 27.8 : null,
                current_l3: isTri ? 29.1 : null,
                power_w_total: isTri ? (5500 + Math.random() * 4000) : (800 + Math.random() * 500),
                power_factor: isTri ? 0.92 : 0.96,
                frequency_hz: 50.01,
                energy_kwh_total: isTri ? 12500 : 2100,
            };

            await client.query(
                `INSERT INTO telemetry_realtime (device_id, device_type, data, last_updated)
                 VALUES ($1, $2, $3, NOW())
                 ON CONFLICT (device_id)
                 DO UPDATE SET data = $3, device_type = $2, last_updated = NOW()`,
                [device.id, device.device_type, JSON.stringify(latestData)]
            );

            console.log(`   ✓ ${device.name}: ${deviceReadings} readings`);
        }

        await client.query('COMMIT');
        console.log(`\n[4] ✅ Done! ${totalInserted} total telemetry readings inserted.`);
        console.log(`    Open http://localhost:5173 and navigate to the factory dashboard!`);
    } catch (err) {
        await client.query('ROLLBACK');
        console.error('Error:', err.message);
    } finally {
        client.release();
        process.exit(0);
    }
}

main();

File: scripts/simulate_telemetry.py
#!/usr/bin/env python3
"""
FPSaver — Create devices + Simulate telemetry data
Creates 4 devices in the factory, then inserts realistic telemetry data.
"""
import requests
import json
import random
import math
import sys
from datetime import datetime, timedelta

BASE = "http://localhost:3000/api"

# Login as SuperAdmin
print("[1] Logging in as SuperAdmin...")
r = requests.post(f"{BASE}/auth/login", json={"email": "admin@fpsaver.com", "password": "SuperAdmin2026!"})
token = r.json()["data"]["accessToken"]
headers = {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}

# Get factory
print("[2] Getting factory...")
r = requests.get(f"{BASE}/superadmin/factories", headers=headers)
factories = r.json()["data"]
if not factories:
    print("   No factories found! Create one first.")
    sys.exit(1)
factory = factories[0]
factory_id = factory["id"]
print(f"   Factory: {factory['name']} ({factory_id})")

# Check existing devices
r = requests.get(f"{BASE}/factories/{factory_id}/devices", headers=headers)
existing_devices = r.json()["data"]
print(f"   Existing devices: {len(existing_devices)}")

# Create devices if none exist
if len(existing_devices) == 0:
    print("[3] Creating devices...")
    devices_to_create = [
        {"name": "Compresor 1", "device_type": "trifasica", "modbus_address": 1, "model": "EM340", "serial_number": "SN-CG-001"},
        {"name": "Compresor 2", "device_type": "trifasica", "modbus_address": 2, "model": "EM340", "serial_number": "SN-CG-002"},
        {"name": "Horno Industrial", "device_type": "trifasica", "modbus_address": 3, "model": "EM340", "serial_number": "SN-CG-003"},
        {"name": "Iluminación Nave", "device_type": "monofasica", "modbus_address": 4, "model": "EM111", "serial_number": "SN-CG-004"},
    ]
    for d in devices_to_create:
        r = requests.post(f"{BASE}/factories/{factory_id}/devices", headers=headers, json=d)
        if r.status_code == 201:
            print(f"   ✓ Created: {d['name']} ({r.json()['data']['id'][:8]}...)")
        else:
            print(f"   ✗ Failed: {d['name']} - {r.text[:100]}")

    # Re-fetch devices  
    r = requests.get(f"{BASE}/factories/{factory_id}/devices", headers=headers)
    existing_devices = r.json()["data"]
else:
    print("[3] Devices already exist, skipping creation.")

devices = existing_devices
print(f"   Total devices: {len(devices)}")

# Simulate telemetry data for the past 48 hours
print("[4] Generating simulated telemetry (48h of data)...")

def generate_trifasica_reading(hour_of_day, noise=0.1):
    """Generate realistic 3-phase meter readings based on time of day"""
    # Industrial load profile: higher during work hours
    if 6 <= hour_of_day < 22:
        load_factor = 0.6 + 0.3 * math.sin(math.pi * (hour_of_day - 6) / 16)
    else:
        load_factor = 0.15 + 0.05 * random.random()
    
    base_power = random.uniform(3000, 15000) * load_factor
    
    # Voltages (230V L-N nominal for each phase)
    v1 = 228 + random.gauss(0, 2)
    v2 = 229 + random.gauss(0, 2)
    v3 = 231 + random.gauss(0, 2)
    
    # Currents
    i1 = base_power / (v1 * 0.85) * (1 + random.gauss(0, noise))
    i2 = base_power / (v2 * 0.87) * (1 + random.gauss(0, noise))
    i3 = base_power / (v3 * 0.83) * (1 + random.gauss(0, noise))
    
    # Power factors
    pf1 = min(0.99, max(0.75, 0.92 + random.gauss(0, 0.03)))
    pf2 = min(0.99, max(0.75, 0.91 + random.gauss(0, 0.03)))
    pf3 = min(0.99, max(0.75, 0.90 + random.gauss(0, 0.03)))
    
    # Active power per phase
    p1 = v1 * i1 * pf1
    p2 = v2 * i2 * pf2
    p3 = v3 * i3 * pf3
    
    return {
        "voltage_l1_n": round(v1, 1), "voltage_l2_n": round(v2, 1), "voltage_l3_n": round(v3, 1),
        "current_l1": round(i1, 2), "current_l2": round(i2, 2), "current_l3": round(i3, 2),
        "power_w_l1": round(p1, 1), "power_w_l2": round(p2, 1), "power_w_l3": round(p3, 1),
        "power_w_total": round(p1 + p2 + p3, 1),
        "power_va_l1": round(v1 * i1, 1), "power_va_l2": round(v2 * i2, 1), "power_va_l3": round(v3 * i3, 1),
        "power_var_l1": round(v1 * i1 * math.sin(math.acos(pf1)), 1),
        "power_var_l2": round(v2 * i2 * math.sin(math.acos(pf2)), 1),
        "power_var_l3": round(v3 * i3 * math.sin(math.acos(pf3)), 1),
        "power_factor_l1": round(pf1, 3), "power_factor_l2": round(pf2, 3), "power_factor_l3": round(pf3, 3),
        "power_factor": round((pf1 + pf2 + pf3) / 3, 3),
        "frequency_hz": round(50 + random.gauss(0, 0.05), 2),
        "energy_kwh_total": round(random.uniform(1000, 50000), 1),
    }

def generate_monofasica_reading(hour_of_day):
    """Generate realistic single-phase meter readings"""
    if 6 <= hour_of_day < 22:
        load_factor = 0.5 + 0.4 * math.sin(math.pi * (hour_of_day - 6) / 16)
    else:
        load_factor = 0.1
    
    voltage = 230 + random.gauss(0, 2)
    power = random.uniform(500, 3000) * load_factor
    pf = min(0.99, max(0.8, 0.95 + random.gauss(0, 0.02)))
    current = power / (voltage * pf)
    
    return {
        "voltage": round(voltage, 1),
        "current": round(current, 2),
        "power_w": round(power, 1),
        "power_va": round(voltage * current, 1),
        "power_var": round(voltage * current * math.sin(math.acos(pf)), 1),
        "power_factor": round(pf, 3),
        "frequency_hz": round(50 + random.gauss(0, 0.05), 2),
        "energy_kwh_total": round(random.uniform(100, 5000), 1),
    }

# Insert telemetry via REST endpoint
now = datetime.utcnow()
total_readings = 0
batch_size = 20

for device in devices:
    device_id = device["id"]
    device_type = device["device_type"]
    print(f"   Generating data for {device['name']} ({device_type})...")
    
    readings = []
    for h in range(48 * 12):  # Every 5 minutes for 48 hours = 576 readings
        ts = now - timedelta(minutes=h * 5)
        hour_of_day = ts.hour
        
        if device_type == "trifasica":
            data = generate_trifasica_reading(hour_of_day)
        else:
            data = generate_monofasica_reading(hour_of_day)
        
        readings.append({
            "factory_id": factory_id,
            "device_id": device_id,
            "device_type": device_type,
            "timestamp": ts.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "data": data,
        })
    
    # Insert in batches
    for i in range(0, len(readings), batch_size):
        batch = readings[i:i + batch_size]
        r = requests.post(f"{BASE}/telemetry/batch", headers=headers, json={"readings": batch})
        if r.status_code != 201:
            print(f"      ✗ Batch {i//batch_size} failed: {r.text[:80]}")
            break
        total_readings += len(batch)
    
    print(f"      ✓ {len(readings)} readings inserted")

print(f"\n[5] Done! {total_readings} total telemetry readings inserted.")
print(f"    Factory: {factory['name']}")
print(f"    Devices: {len(devices)}")
print(f"    Period: {(now - timedelta(hours=48)).strftime('%Y-%m-%d %H:%M')} → {now.strftime('%Y-%m-%d %H:%M')}")
print(f"\n    Now login at http://localhost:5173 and check the factory dashboard!")
